File: array_and_hashing/TopKFrequentElements.java
import java.util.*;

public class TopKFrequentElements {

    public static int[] topKFrequentHeap(int[] nums, int k) {
        Map<Integer, Integer> count = new HashMap<>();
        for (int num : nums) count.merge(num, 1, Integer::sum);

        // Min-heap of size k, ordered by frequency
        PriorityQueue<Integer> heap = new PriorityQueue<>(Comparator.comparingInt(count::get));
        for (int num : count.keySet()) {
            heap.offer(num);
            if (heap.size() > k) heap.poll();
        }

        int[] result = new int[k];
        for (int i = k - 1; i >= 0; i--) result[i] = heap.poll();
        return result;
    }

    public static int[] topKFrequentBucketSort(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();
        for (int num : nums) freq.merge(num, 1, Integer::sum);

        // Create buckets where index = frequency
        @SuppressWarnings("unchecked")
        List<Integer>[] buckets = new List[nums.length + 1];
        for (int i = 0; i <= nums.length; i++) buckets[i] = new ArrayList<>();
        for (Map.Entry<Integer, Integer> e : freq.entrySet()) buckets[e.getValue()].add(e.getKey());

        // Collect top k elements from highest frequency bucket down
        int[] result = new int[k];
        int idx = 0;
        for (int i = buckets.length - 1; i > 0 && idx < k; i--) {
            for (int num : buckets[i]) {
                result[idx++] = num;
                if (idx == k) return result;
            }
        }
        return result;
    }
}

File: array_and_hashing/top_k_frequent_elements.ts
function topKFrequentSort(nums: number[], k: number): number[] {
  const count = new Map<number, number>();
  for (const num of nums) count.set(num, (count.get(num) ?? 0) + 1);

  return [...count.keys()]
    .sort((a, b) => count.get(b)! - count.get(a)!)
    .slice(0, k);
}

function topKFrequentBucketSort(nums: number[], k: number): number[] {
  const freq = new Map<number, number>();
  for (const num of nums) freq.set(num, (freq.get(num) ?? 0) + 1);

  // Create buckets where index = frequency
  const buckets: number[][] = Array.from({ length: nums.length + 1 }, () => []);
  for (const [num, count] of freq) buckets[count].push(num);

  // Collect top k elements from highest frequency bucket down
  const result: number[] = [];
  for (let i = buckets.length - 1; i > 0 && result.length < k; i--) {
    for (const num of buckets[i]) {
      result.push(num);
      if (result.length === k) return result;
    }
  }
  return result;
}

function topKFrequentQuickSelect(nums: number[], k: number): number[] {
  const count = new Map<number, number>();
  for (const num of nums) count.set(num, (count.get(num) ?? 0) + 1);

  const unique = [...count.keys()];
  const n = unique.length;

  if (k >= n) return unique;

  function partition(left: number, right: number, pivotIndex: number): number {
    const pivotFreq = count.get(unique[pivotIndex])!;
    [unique[pivotIndex], unique[right]] = [unique[right], unique[pivotIndex]];

    let storeIndex = left;
    for (let i = left; i < right; i++) {
      if (count.get(unique[i])! < pivotFreq) {
        [unique[i], unique[storeIndex]] = [unique[storeIndex], unique[i]];
        storeIndex++;
      }
    }
    [unique[right], unique[storeIndex]] = [unique[storeIndex], unique[right]];
    return storeIndex;
  }

  function quickSelect(left: number, right: number, kSmallest: number): void {
    if (left >= right) return;
    const pivotIndex = left + Math.floor(Math.random() * (right - left + 1));
    const finalIndex = partition(left, right, pivotIndex);

    if (finalIndex === kSmallest) return;
    else if (finalIndex > kSmallest)
      quickSelect(left, finalIndex - 1, kSmallest);
    else quickSelect(finalIndex + 1, right, kSmallest);
  }

  quickSelect(0, n - 1, n - k);
  return unique.slice(n - k);
}
